File: packages/provide-uterm-go/shell/output.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package shell

import (
	"strings"
	"unicode/utf8"
)

// ANSI escape constants. Ported verbatim from _output.py. All multi-line
// strings use CRLF line endings (terminal protocol).
const (
	Reset   = "\x1b[0m"
	Bold    = "\x1b[1m"
	Dim     = "\x1b[2m"
	Green   = "\x1b[32m"
	Yellow  = "\x1b[33m"
	Red     = "\x1b[31m"
	Cyan    = "\x1b[36m"
	Blue    = "\x1b[34m"
	Magenta = "\x1b[35m"

	// ClearScreen erases the screen and homes the cursor.
	ClearScreen = "\x1b[2J\x1b[H"
)

// Prompt is the ushell command prompt.
var Prompt = Green + "❯" + Reset + " "

// Banner is the ushell startup banner.
var Banner = Bold + Cyan + "ushell" + Reset + " " + Dim + "— Python REPL inside your terminal" + Reset + "\r\n" +
	Dim + "Type " + Reset + "help" + Dim + " for available commands." + Reset + "\r\n\r\n"

// ErrorMsg formats a red "error:" line.
func ErrorMsg(text string) string {
	return Red + "error:" + Reset + " " + text + "\r\n"
}

// InfoMsg formats a dim informational line.
func InfoMsg(text string) string {
	return Dim + text + Reset + "\r\n"
}

// SuccessMsg formats a green success line.
func SuccessMsg(text string) string {
	return Green + text + Reset + "\r\n"
}

// Heading formats a bold-cyan heading line.
func Heading(text string) string {
	return Bold + Cyan + text + Reset + "\r\n"
}

// FmtKV formats a key/value pair with the key left-padded to width columns.
// Mirrors _output.fmt_kv (default width 20).
func FmtKV(key, value string, width int) string {
	return "  " + Dim + padRight(key, width) + Reset + value + "\r\n"
}

// FmtKVDefault is FmtKV with the default width of 20.
func FmtKVDefault(key, value string) string {
	return FmtKV(key, value, 20)
}

// FmtTable formats rows as a fixed-width table. When headers is non-nil a bold
// header row plus a dashed separator precede the data. Mirrors
// _output.fmt_table.
func FmtTable(rows [][]string, headers []string) string {
	if len(rows) == 0 {
		return InfoMsg("(no results)")
	}
	// Per-column max width over the data cells.
	ncols := 0
	for _, r := range rows {
		if len(r) > ncols {
			ncols = len(r)
		}
	}
	widths := make([]int, ncols)
	for _, r := range rows {
		for i, cell := range r {
			if w := utf8.RuneCountInString(cell); w > widths[i] {
				widths[i] = w
			}
		}
	}
	if headers != nil {
		for i := 0; i < len(headers) && i < ncols; i++ {
			if w := utf8.RuneCountInString(headers[i]); w > widths[i] {
				widths[i] = w
			}
		}
	}

	var lines []string
	if headers != nil {
		cells := make([]string, 0, len(headers))
		for i, h := range headers {
			w := 0
			if i < len(widths) {
				w = widths[i]
			}
			cells = append(cells, Bold+padRight(h, w)+Reset)
		}
		lines = append(lines, "  "+strings.Join(cells, "  "))
		dashes := make([]string, len(widths))
		for i, w := range widths {
			dashes[i] = strings.Repeat("-", w)
		}
		lines = append(lines, "  "+strings.Join(dashes, "  "))
	}
	for _, r := range rows {
		cells := make([]string, len(r))
		for i, cell := range r {
			w := 0
			if i < len(widths) {
				w = widths[i]
			}
			cells[i] = padRight(cell, w)
		}
		lines = append(lines, "  "+strings.Join(cells, "  "))
	}
	return strings.Join(lines, "\r\n") + "\r\n"
}

// padRight left-justifies s to w display columns (rune count), matching
// Python's f"{s:<{w}}" for the ASCII cells used here.
func padRight(s string, w int) string {
	n := utf8.RuneCountInString(s)
	if n >= w {
		return s
	}
	return s + strings.Repeat(" ", w-n)
}

File: packages/provide-uterm-csharp/src/Provide.Uterm/Shell/Output.cs
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

using System.Text;

namespace Provide.Uterm.Shell;

/// <summary>ANSI helpers. Port of packages/provide-uterm-go/shell/output.go (byte-aligned with Python).</summary>
public static class ShellOutput
{
    public const string Reset = "\x1b[0m";
    public const string Bold = "\x1b[1m";
    public const string Dim = "\x1b[2m";
    public const string Green = "\x1b[32m";
    public const string Yellow = "\x1b[33m";
    public const string Red = "\x1b[31m";
    public const string Cyan = "\x1b[36m";
    public const string Blue = "\x1b[34m";
    public const string Magenta = "\x1b[35m";
    public const string ClearScreen = "\x1b[2J\x1b[H";

    public static readonly string Prompt = Green + "❯" + Reset + " ";
    public static readonly string Banner =
        Bold + Cyan + "ushell" + Reset + " " + Dim + "— Python REPL inside your terminal" + Reset + "\r\n" +
        Dim + "Type " + Reset + "help" + Dim + " for available commands." + Reset + "\r\n\r\n";

    public static string ErrorMsg(string text) => Red + "error:" + Reset + " " + text + "\r\n";
    public static string InfoMsg(string text) => Dim + text + Reset + "\r\n";
    public static string SuccessMsg(string text) => Green + text + Reset + "\r\n";
    public static string Heading(string text) => Bold + Cyan + text + Reset + "\r\n";

    public static string FmtKv(string key, string value, int width = 20) =>
        "  " + Dim + PadRight(key, width) + Reset + value + "\r\n";

    public static string FmtKvDefault(string key, string value) => FmtKv(key, value, 20);

    /// <summary>Legacy dictionary dump used by older call sites.</summary>
    public static string FmtKv(IReadOnlyDictionary<string, string> kv)
    {
        var sb = new StringBuilder();
        foreach (var (k, v) in kv)
        {
            sb.Append(FmtKvDefault(k, v));
        }

        return sb.ToString();
    }

    public static string FmtTable(IReadOnlyList<IReadOnlyList<string>> rows, IReadOnlyList<string>? headers)
    {
        if (rows.Count == 0)
        {
            return InfoMsg("(no results)");
        }

        var ncols = 0;
        foreach (var r in rows)
        {
            if (r.Count > ncols)
            {
                ncols = r.Count;
            }
        }

        var widths = new int[ncols];
        foreach (var r in rows)
        {
            for (var i = 0; i < r.Count; i++)
            {
                var w = r[i].Length; // ASCII table cells like Go's rune count for ASCII
                if (w > widths[i])
                {
                    widths[i] = w;
                }
            }
        }

        if (headers is not null)
        {
            for (var i = 0; i < headers.Count && i < ncols; i++)
            {
                var w = headers[i].Length;
                if (w > widths[i])
                {
                    widths[i] = w;
                }
            }
        }

        var lines = new List<string>();
        if (headers is not null)
        {
            var cells = new List<string>();
            for (var i = 0; i < headers.Count; i++)
            {
                var w = i < widths.Length ? widths[i] : 0;
                cells.Add(Bold + PadRight(headers[i], w) + Reset);
            }

            lines.Add("  " + string.Join("  ", cells));
            var dashes = widths.Select(w => new string('-', w)).ToArray();
            lines.Add("  " + string.Join("  ", dashes));
        }

        foreach (var r in rows)
        {
            var cells = new string[r.Count];
            for (var i = 0; i < r.Count; i++)
            {
                var w = i < widths.Length ? widths[i] : 0;
                cells[i] = PadRight(r[i], w);
            }

            lines.Add("  " + string.Join("  ", cells));
        }

        return string.Join("\r\n", lines) + "\r\n";
    }

    public static string PadRight(string s, int w)
    {
        if (s.Length >= w)
        {
            return s;
        }

        return s + new string(' ', w - s.Length);
    }
}
